src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -212,5 +214,22 @@ public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions
                     break;
             }
         }
+
+        private string FormatEnumValue(string value)
+        {
+            if (value.IndexOf(EnumValueSeparator) == -1)
+            {
+                return _namingPolicy.ConvertName(value);
+            }
+
+            string[] enumValues = value.Split(EnumValueSeparator);
```

Comment:
Is this perf sensitive? Seems a shame to create an array and temporary strings.

I guess this needs https://github.com/dotnet/runtime/pull/295 (maybe you'd like to volunteer to review that)

Reply:
Not wrt. throughput. It only affects first-time serialization of the enum value, since the conversion is cached.

> I guess this needs #295 

Yes I had a local implementation of the split method. There is a bit of a perf boost - less allocs and less CPU time, but I decided not to block the fix here (to get all the changes where we are changing serialized output front-loaded). I did make a note-to-self to drive that PR in, and I'll put a to-do in this PR.

(BTW) Another thing to do here is to cache the result of the transformation here as a `JsonEncodedText` which is faster for the writer to write (vs a string as we do today). I'll push a commit to cache encoded strings since that is easy to do in this PR.


|                           Method |            enumValue |       Mean |      Error |     StdDev |     Median | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |--------------------- |-----------:|-----------:|-----------:|-----------:|------:|--------:|-------:|------:|------:|----------:|
|              StringSplitThenJoin |           Compressed | 135.136 ns |  7.2528 ns | 19.9763 ns | 124.516 ns |  1.00 |    0.00 | 0.0210 |     - |     - |      88 B |
|             SplanSplitThenCopyTo |           Compressed | 134.745 ns |  3.7377 ns | 10.4192 ns | 131.676 ns |  1.01 |    0.15 | 0.0210 |     - |     - |      88 B |
|                                  |                      |            |            |            |            |       |         |        |       |       |           |
|              StringSplitThenJoin | Direc(...)tream [38] | 405.194 ns | 11.5382 ns | 33.2904 ns | 392.147 ns |  1.00 |    0.00 | 0.1450 |     - |     - |     608 B |
|             SplanSplitThenCopyTo | Direc(...)tream [38] | 364.852 ns |  0.6745 ns |  0.5632 ns | 364.704 ns |  0.87 |    0.06 | 0.1087 |     - |     - |     456 B |
|                                  |                      |            |            |            |            |       |         |        |       |       |           |
|              StringSplitThenJoin |  Read(...)Data [100] | 779.538 ns |  2.6841 ns |  2.3794 ns | 779.540 ns |  1.00 |    0.00 | 0.3519 |     - |     - |    1472 B |
|             SplanSplitThenCopyTo |  Read(...)Data [100] | 739.864 ns |  3.9898 ns |  3.3316 ns | 737.661 ns |  0.95 |    0.00 | 0.2747 |     - |     - |    1152 B |

<details>
<summary>Benchmark code with ROS split implementation (click to view)</summary>

```cs
using BenchmarkDotNet.Attributes;
using System;
using System.Buffers;
using System.IO;
using System.Text.Json;

namespace MiscBenchmarks
{
    [MemoryDiagnoser]
    public class EnumFlagNamingPolicy
    {
        [Params(
            FileAttributes.Compressed,
            FileAttributes.Directory | FileAttributes.Compressed | FileAttributes.IntegrityStream,
            FileAttributes.Hidden | FileAttributes.Normal | FileAttributes.ReadOnly | FileAttributes.NoScrubData | FileAttributes.SparseFile |
            FileAttributes.Archive | FileAttributes.Encrypted | FileAttributes.Compressed | FileAttributes.NotContentIndexed
            )]
        public FileAttributes enumValue;

        private string enumAsStr;

        private static readonly JsonNamingPolicy s_namingPolicy = JsonNamingPolicy.CamelCase;

        private const string Separator = ", ";

        [GlobalSetup]
        public void Setup()
        {
            enumAsStr = enumValue.ToString();
        }

        [Benchmark(Baseline = true)]
        public JsonEncodedText StringSplitThenJoin()
        {
            if (!enumAsStr.Contains(Separator))
            {
                return JsonEncodedText.Encode(enumAsStr);
            }

            string[] enumValues = enumAsStr.Split(Separator);

            for (int i = 0; i < enumValues.Length; i++)
            {
                enumValues[i] = s_namingPolicy.ConvertName(enumValues[i]);
            }

            return JsonEncodedText.Encode(string.Join(", ", enumValues));
        }

        [Benchmark]
        public JsonEncodedText SplanSplitThenCopyTo()
        {
            if (!enumAsStr.Contains(Separator))
            {
                return JsonEncodedText.Encode(enumAsStr);
            }

            ReadOnlySpan<char> span = enumAsStr.AsSpan();
            SpanSplitSequenceEnumerator<char> enumerator = Split(span, Separator);

            char[] arr;
            Span<char> converted = arr = ArrayPool<char>.Shared.Rent(enumAsStr.Length * 2);

            ReadOnlySpan<char> separator = Separator;
            int builderPos = 0;

            foreach (Range range in enumerator)
            {
                int start = range.Start.Value;
                int length = range.End.Value - start;

                string input = new string(span.Slice(range.Start.Value, length));
                string output = s_namingPolicy.ConvertName(input);

                if (builderPos + output.Length + Separator.Length >= converted.Length)
                {
                    char[] newArr = ArrayPool<char>.Shared.Rent(converted.Length * 2);
                    converted.CopyTo(newArr);

                    ArrayPool<char>.Shared.Return(arr);
                    converted = arr = newArr;
                }

                ((ReadOnlySpan<char>)output).CopyTo(converted.Slice(builderPos));
                separator.CopyTo(converted.Slice(builderPos + output.Length));

                builderPos += output.Length + Separator.Length;
            }

            JsonEncodedText val = JsonEncodedText.Encode(converted.Slice(0, builderPos - Separator.Length));

            ArrayPool<char>.Shared.Return(arr);

            return val;
        }

        internal static SpanSplitSequenceEnumerator<T> Split<T>(ReadOnlySpan<T> span, ReadOnlySpan<T> separator) where T : IEquatable<T>
        => new SpanSplitSequenceEnumerator<T>(span, separator);

        internal ref struct SpanSplitSequenceEnumerator<T> where T : IEquatable<T>
        {
            private readonly ReadOnlySpan<T> _buffer;
            private readonly ReadOnlySpan<T> _separator;

            private readonly int _separatorLen;
            private int _offset;

            private int _start;
            private int _end;

            public SpanSplitSequenceEnumerator<T> GetEnumerator() => this;

            public SpanSplitSequenceEnumerator(ReadOnlySpan<T> span, ReadOnlySpan<T> separator)
            {
                _buffer = span;
                _separator = separator;
                _separatorLen = separator.Length;
                _start = 0;
                _end = 0;
                _offset = 0;
            }

            public Range Current => new Range(_start, _end);

            public bool MoveNext()
            {
                if (_offset > _buffer.Length)
                {
                    return false;
                }

                ReadOnlySpan<T> slice = _buffer.Slice(_offset);
                _start = _offset;

                int separatorIndex = slice.IndexOf(_separator);
                
                int subStringLength = separatorIndex == -1
                    ? _buffer.Length - _start
                    : separatorIndex;

                _end = _start + subStringLength;
                _offset = _end + _separatorLen;

                return true;
            }
        }
    }
}
```

</details>

Reply:
Look at the guy with the Benchmark.NET results! 😺 

> I did make a note-to-self to drive that PR in, and I'll put a to-do in this PR.

Love it.

Reply:
😎